File: locus-api-android/src/main/java/locus/api/android/utils/UtilsBitmap.java
package locus.api.android.utils;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

import locus.api.utils.DataReaderBigEndian;
import locus.api.utils.DataWriterBigEndian;
import locus.api.utils.Logger;

/**
 * Created by menion on 09/10/2015.
 * This code is part of Locus project from Asamm Software, s. r. o.
 */
public class UtilsBitmap {

    // tag for logger
    private static final String TAG = "UtilsBitmap";

    public static Bitmap readBitmap(DataReaderBigEndian dr) {
        int size = 0;
        if ((size = dr.readInt()) > 0) {
            byte[] data = dr.readBytes(size);
            return getBitmap(data);
        } else {
            return null;
        }
    }

    public static void writeBitmap(DataWriterBigEndian dw, Bitmap bitmap,
                                   Bitmap.CompressFormat format) throws IOException {
        if (bitmap == null) {
            dw.writeInt(0);
        } else {
            byte[] data = getBitmap(bitmap, format);
            if (data == null || data.length == 0) {
                Logger.logW(TAG, "writeBitmap(), unknown problem");
                dw.writeInt(0);
            } else {
                dw.writeInt(data.length);
                dw.write(data);
            }
        }
    }

    public static byte[] getBitmap(Bitmap bitmap, Bitmap.CompressFormat format) {
        ByteArrayOutputStream baos = null;
        try {
            baos = new ByteArrayOutputStream();
            // create compressed byte array. Value 80 is used for JPEG. PNG format ignore this value
            if (bitmap.compress(format, 80, baos)) {
                return baos.toByteArray();
            } else {
                Logger.logW(TAG, "Problem with converting image to byte[]");
                return null;
            }
        } catch (Exception e) {
            Logger.logE(TAG, "getBitmap(" + bitmap + ")", e);
            return null;
        } finally {
            locus.api.utils.Utils.closeStream(baos);
        }
    }

    public static Bitmap getBitmap(byte[] data) {
        try {
            return BitmapFactory.decodeByteArray(data, 0, data.length);
        } catch (Exception e) {
            Logger.logE(TAG, "getBitmap(" + data + ")", e);
            return null;
        }
    }
}

File: locus-api-android/src/main/java/locus/api/android/utils/UtilsBitmap.kt
/**
 * Created by menion on 09/10/2015.
 * This code is part of Locus project from Asamm Software, s. r. o.
 */
package locus.api.android.utils

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.asamm.loggerV2.logE
import com.asamm.loggerV2.logW
import locus.api.utils.DataReaderBigEndian
import locus.api.utils.DataWriterBigEndian
import java.io.ByteArrayOutputStream

object UtilsBitmap {

    // tag for logger
    private const val TAG = "UtilsBitmap"

    fun readBitmap(dr: DataReaderBigEndian): Bitmap? {
        val size: Int = dr.readInt()
        return if (size > 0) {
            getBitmap(dr.readBytes(size))
        } else {
            null
        }
    }

    fun writeBitmap(
        dw: DataWriterBigEndian, bitmap: Bitmap?,
        format: Bitmap.CompressFormat
    ) {
        if (bitmap == null) {
            dw.writeInt(0)
        } else {
            val data = getBitmap(bitmap, format)
            if (data == null || data.isEmpty()) {
                logW(tag = TAG) { "writeBitmap(), unknown problem" }
                dw.writeInt(0)
            } else {
                dw.writeInt(data.size)
                dw.write(data)
            }
        }
    }

    fun getBitmap(bitmap: Bitmap, format: Bitmap.CompressFormat): ByteArray? {
        var baos: ByteArrayOutputStream? = null
        return try {
            baos = ByteArrayOutputStream()
            // create compressed byte array. Value 80 is used for JPEG. PNG format ignore this value
            if (bitmap.compress(format, 80, baos)) {
                baos.toByteArray()
            } else {
                logW(tag = TAG) { "Problem with converting image to byte[]" }
                null
            }
        } catch (e: Exception) {
            logE(tag = TAG, ex = e) { "getBitmap($bitmap)" }
            null
        } finally {
            locus.api.utils.Utils.closeStream(baos)
        }
    }

    fun getBitmap(data: ByteArray): Bitmap? {
        return try {
            BitmapFactory.decodeByteArray(data, 0, data.size)
        } catch (e: Exception) {
            logE(tag = TAG, ex = e) { "getBitmap($data)" }
            null
        }
    }
}
